### PR TITLE
HDF5 update

### PR DESCRIPTION
**Pull request by dzenanz**
Migrated from [Gerrit](http://review.source.kitware.com/#/c/23844/) code review system.

HDF5 2018-10-15 (545c5fb2). Code extracted from:
https://bitbucket.hdfgroup.org/scm/hdffv/hdf5.git
at commit 545c5fb2a957c245f283c6db5c429b7b89cda622 (branch: hdf5_1_10_4).

### CODE REVIEW

**Comment by thewtex** [comment]
The mac build is failing with:

```
2018-11-06T14:34:56.5612690Z CMake Error: Attempt to add a custom rule to output "/Users/vsts/agent/2.141.1/work/1/s-build/Modules/ThirdParty/HDF5/src/itkhdf5/shared/H5Tinit.c.rule" which already has a custom rule.
2018-11-06T14:34:56.5613330Z CMake Error: Attempt to add a custom rule to output "/Users/vsts/agent/2.141.1/work/1/s-build/Modules/ThirdParty/HDF5/src/itkhdf5/shared/H5lib_settings.c.rule" which already has a custom rule.
```

The *ghostflow-check-master* is a false positive.

File: Modules/ThirdParty/HDF5/src/itkhdf5/src/CMakeLists.txt [hunk 1]
```diff
@@ -872,10 +884,36 @@ add_custom_command (
     COMMAND ${CMAKE_CROSSCOMPILING_EMULATOR} $<TARGET_FILE:H5detect>
     ARGS > ${HDF5_BINARY_DIR}/H5Tinit.c
     DEPENDS ${INTERNAL_BUILD_PREFIX}H5detect
-)
+    WORKING_DIRECTORY ${HDF5_GENERATED_SOURCE_DIR}
+  )
+  if (BUILD_SHARED_LIBS)
+    if (NOT EXISTS ${HDF5_GENERATED_SOURCE_DIR}/shared/H5Tinit.c)
+      add_custom_command (
+          OUTPUT ${HDF5_GENERATED_SOURCE_DIR}/shared/H5Tinit.c
```
Comment (bradking) [inline Modules/ThirdParty/HDF5/src/itkhdf5/src/CMakeLists.txt]: This `CMakeLists.txt` file already has a similar rule:

```cmake
  add_custom_command (
      OUTPUT ${HDF5_BINARY_DIR}/shared/H5Tinit.c
```

This is the cause of the CMake Error about duplicate custom rules.

Comment (dzenanz) [reply Modules/ThirdParty/HDF5/src/itkhdf5/src/CMakeLists.txt]: The error does not occur if HDF5 is built on its own. The problem appears when it is bundled in ITK.

Comment (bradking) [reply Modules/ThirdParty/HDF5/src/itkhdf5/src/CMakeLists.txt]: I suspect that ITK's modifications cause the duplicate code, perhaps due to incorrect conflict resolution.

**Comment by hjmjohnson** [comment]
Superceeded by #292

**Pull request closed without merging**